### ISSUE TITLE
libtorch.so file size is very large

### ISSUE THREAD

**Issue by boulabiar** [pytorch/pytorch#34058]
## 🐛 Bug

The library size of torch is very large (~267MB without CUDA, 1.2 GB with CUDA).
This can be caused by using a library like Intel IPP similar to this opencv bug.
https://github.com/opencv/opencv/issues/15177
This limits the use of pytorch models in small environments (without CUDA).

## To Reproduce

1. Download the libtorch C++ library

## Expected behavior

Library size without CUDA should be a lot less than that. or at least providing libtorch_tiny or ways to create it.

## Environment
 - PyTorch Version: 1.4
 - OS: Linux

**Comment by ezyang** [comment]
cc @soumith marking high priority to at least respond to this issue; this might be non-actionable

**Comment by boulabiar** [comment]
I have problems deploying to C++, conversion to onnx is not always direct for non standard models (and to Caffe2 you need to go through onnx), and with a scripted model, the library size is very large.
Getting it at least below 150MB is already a big advantage, and as I said, this can be done deactivating a flag in the build or providing a pytorch_tiny mode.

**Comment by yf225** [comment]
This might not be fixable. We will investigate and then give a response on this.

**Comment by boulabiar** [comment]
I hope you find something, or matching size taken by each component to those 267MB.
it can be a single library "inline"ing all functions for avx, avx2, avx512, ... or similar like the case for opencv.

**Comment by honglh** [comment]
I made a build with USE_CUDA=0 and it seems the library size is small now 

```
-rw-r--r-- 1 1000 1000      7480 Apr  6 06:00 libtorch.so
-rw-r--r-- 1 1000 1000 118034304 Apr  6 06:00 libtorch_cpu.so

```

**Comment by igorbrigadir** [comment]
If you do need CUDA, another option to drastically reduce the binary size is to limit the supported CUDA versions to the one you need - if you know exactly what kind of hardware you have, eg, setting `TORCH_CUDA_ARCH_LIST="7.0+PTX"` would only support latest CUDA cards - see Compute Capability and cuda versions and cards in a list here: https://en.wikipedia.org/wiki/CUDA#GPUs_supported

Other ways to reduce the binary size is to exclude other things you may not need - like `USE_NCCL=0` if you don't need multi gpu support, or CUDNN with `USE_CUDNN=0` if you can get away with not having CUDNN support.

Getting `TORCH_CUDA_ARCH_LIST` down to the one you need makes the most difference.

**Comment by boulabiar** [comment]
I don't need CUDA, so being able to reduce it to 120MB is very convenient, and if it is the case why the libtorch cpu package is big? Does it still have CUDA symbols in it ?
This one:
https://download.pytorch.org/libtorch/cpu/libtorch-cxx11-abi-shared-with-deps-1.4.0%2Bcpu.zip

**Comment by igorbrigadir** [comment]
Ah a big chunk of that would have to be MKL. To disable that you can uninstall mkl when building (replace mkl with `nomkl` too if using the miniconda installer like in https://github.com/pytorch/pytorch/blob/master/docker/pytorch/Dockerfile#L19 and set `USE_MKLDNN=0` but you'll also need to set `BLAS=Eigen` because you need MKL for BLAS too, unless you use Eigen. This is one of the ways i got it to compile in the past, not quite sure what kind of feature / performance hit this will have.

**Comment by rilut** [comment]
yes, 234M for libtorch_cpu.dylib on torch 1.7.0. It's so hard to stuff pytorch on aliyun function compute

**Comment by Andrej-sens** [comment]
Is there a way to reduce a build with CUDA by configuring it only for inference?

**Comment by BrainSlugs83** [comment]
Any update on this? -- It seems to be a cross platform issue, as on Windows, the sizes are **insane** for the equivalent DLLs:
| File | Size |
| ----------- | ----------- |
| **torch_cuda_cu.dll** | **1.2 GB** |
| cudnn_cnn_infer64_8.dll |  652 MB |

And there are several additional DLLs that are all around the ~200 MB mark...

**Comment by Skylion007** [comment]
With PyTorch 2.0 upon us, and the fact that the minimum supported CUDA is now 11.2 which supports LTO https://discuss.pytorch.org/t/link-time-optimization-support/142005, any plans on revisiting enabling Link Time Optimization for both perf improvements and to shrink the binary size? @ezyang @soumith

**Comment by Skylion007** [comment]
So I just built libtorch_cpu with LTO fairly and enabling it brings binary size of libtorch_cpu.so on master from about 199MB -> 184MB so about a 7-8% reduction in binary size (for free) using gcc-9 and the default ld linker. Definitely added a bit to build time, but works out of the box with no linker errors.

**Comment by Skylion007** [comment]
So full LTO causes issues with some submodules (mainly protobuf), libtorch itself compiles fine though. https://github.com/pytorch/pytorch/pull/93388

**Comment by malfet** [comment]
We probably can build libtorch without ONNX, and that's the only dependency that should be pulling protobuf.

@driazati  tried to enable lto in the past , but it was disabled due to cupti bug, but as we are linking with it dynamically it shouldn't matter.

**Comment by kihot122** [comment]
> So I just built libtorch_cpu with LTO fairly easily and enabling it brings binary size of libtorch_cpu.so on master from about 199MB -> 184MB so about a 7-8% reduction in binary size (for free) using gcc-9 and the default ld linker. Definitely added a bit to build time, but works out of the box with no linker errors.

Could you please provide the exact configuration that you've used? I've been trying to build 2.0 with LTO myself and keep getting assembler errors during the linkage phase.

**Comment by brunoais** [comment]
I was pondering... Is there a way to specify a variation on pip so a more limited GPU variation torch is used?
For example:

`torch[all]` -> a Torch that works in all (basically what we have now)
`torch[Kepler]` -> a Torch that works in Kepler architecture and above
`torch[Maxwell]` -> a Torch that works in Maxwell architecture and above

Unfortunately, I'm not sure how to handle `torch` without brackets in pip if this is done...

With customization which can relate to saved storage space and transfers. Unless you are already spending a lot of money in CI/CD, this is a good option, IMO.

**Comment by make-a** [comment]
> Any update on this? -- It seems to be a cross platform issue, as on Windows, the sizes are **insane** for the equivalent DLLs:
> 
> File	Size
> **torch_cuda_cu.dll**	**1.2 GB**
> cudnn_cnn_infer64_8.dll	652 MB
> And there are several additional DLLs that are all around the ~200 MB mark...

hello, Have you resolved this issue?

**Comment by Skylion007** [comment]
https://github.com/pytorch/pytorch/pull/157791 addresses this by reducing libtorch.so by 40%+ by using better compression of the CUDA code in the binaries.
